File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& a) {
        int n = a.size(); 
        int   ans =0 ;
        
           // time : O(3N) , space :O(n);
          vector<int>prefix(n,0); 
          prefix[0] = a[0] ; 
          for (int i=1 ;i<n ;i++){
              prefix[i] = max(prefix[i-1], a[i]);
          }

          
          vector<int>suffix(n,0); 
          suffix[n- 1] = a[n-1]; 
          for (int i= n- 2;i>=0 ;i--){
              suffix[i] = max(suffix[i+1] , a[i]);
          }

          for (int i=0 ;i<n ;i++){
              ans = ans +  (min(prefix[i] , suffix[i])) -a[i]; 
          }



        return ans ;
        
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, arr):
        """
        :type height: List[int]
        :rtype: int
        """
        #optimised solution ,time : O(n), space :O(1)
        n = len(arr)
        left =0  
        right =n-1 
        leftMax =-1 
        rightMax=-1
        sum=0

        while (left<=right):
            if(arr[left]<=arr[right]):

                if(arr[left]>=leftMax):
                    leftMax = arr[left]
                
                else:
                    #water will be trapped
                    sum += (leftMax-arr[left])
                
                left = left + 1
                
            else :
                if(arr[right]>=rightMax):
                    rightMax=arr[right]
                
                else:
                    #water will trap
                    sum += (rightMax-arr[right])
                
                right = right -1 
        
        return sum 

        